File: ruby/pig-latin/pig_latin.rb
require "strscan"

class PigLatin
  def self.translate(str)
    out = str.split.map do |word|
      s = StringScanner.new(word)

      if s.scan(/[aeiou]|xray|y[^aeiou]/)
        "#{word}ay"
      else
        s.scan(/y/)

        if s.scan_until(/q/)
          s.scan(/u/)
        else
          s.scan(/[^aeiouy]+/)
        end

        "#{s.rest}#{word[0...s.pos]}ay"
      end
    end

    return out.join(" ")
  end
end

File: javascript/pig-latin/pig-latin.js
const CONSONANT = /^(?:s?qu|[bcdfghjklmnpqrstvwz]+|[xy](?=[aeiou]))/

export class translator {
    static translate(str) {
        if (/\s/.test(str))
            return str.split(" ").map(word => this.translate(word)).join(" ");

        let m = str.match(CONSONANT);

        if (m) {
            let length = m[0].length;

            return str.slice(length) + str.slice(0, length) + "ay";
        } else {
            return str + "ay";
        }
    }
}
